### DISCUSSION TITLE
Enabling minimum contrast breaks the color swatches in the theme previewer

### DISCUSSION THREAD

**Discussion by trevortylerlee** [ghostty-org/ghostty#3869] [Q&A]
The actual feature of enforcing a minimum contrast and syntax highlighting still works as expected. It's just the previews for the color swatches that break.
![screenshot 2024-12-29 at 1 11 04 AM](https://github.com/user-attachments/assets/4e403b18-3855-41a7-babb-0761e7a87f79)

**Comment by mitchellh** [comment]
This is working as intended. Their contrast is too low for your setting.

**Reply by trevortylerlee** [reply]
I don't understand... I tried setting minimum-contrast at several values (1.01, 1.1, 3) and every color swatch (for all themes) in the theme previewer is either white or black. How is that working as intended?

**Reply by mitchellh** [reply]
I’ll move this to Q&A and let someone else take a stab

**Reply by mitchellh** [reply]
I wonder if this is actually due to the fg/bg being set to the same with the theme preview and hence the contrast is… 0. That’s something we should look into

**Reply by jcollie** [reply]
That seems a likely possibility, combined with the fact that we're printing solid blocks of color, not foreground-colored text on a colored background.

**Reply by jcollie** [reply]
> I wonder if this is actually due to the fg/bg being set to the same with the theme preview and hence the contrast is… 0. That’s something we should look into

Hmm, no I  just checked the source and the swatches are printed with FG = palette color, BG = background color.

**Reply by jcollie** [reply]
https://github.com/ghostty-org/ghostty/blob/d5703a57e7d7bcf8dd9bd5711d9f351e5a43446b/src/cli/list_themes.zig#L963-L975

**Reply by mitchellh** [reply]
Powerline glyphs are excluded from min contrast. Maybe blocks should be too

**Comment by jcollie** [comment]
`minimum-contrast` works by comparing the brightness of the foreground and background colors and if there isn't a minimum brightness difference it forces the color to all black (#000000) or all white #ffffff. There are a lot of themes that don't have good contrast to start with so they end up looking like what you saw.

Unless you have a severe visual impairment I'd recommend not using `minimum-contrast` and choosing a theme that provides enough contrast "as is".

**Comment by T3sT3ro** [comment]
Would it be possible to provide a similar mode as in the VSCode integrated terminal, i.e. it [increases or decreases foreground luminance until it hits 4.5:1 contrast ratio](https://code.visualstudio.com/docs/terminal/appearance#_minimum-contrast-ratio)? I really like how their terminal has crisp colors and all TUI apps look good out of the box.

Relevant bit to experiment is here: https://github.com/ghostty-org/ghostty/blob/da32534e8a63423d34d46caa9f60f342ed27235f/src/renderer/shaders/cell.v.glsl#L117-L135.

Instead of using WCAG 2 guidelines new APCA / WCAG 3 could be tested instead: https://typefully.com/DanHollick/wcag-3-and-apca-sle13GMW2Brp
(here is the calculator: https://cliambrown.com/contrast/)

**Reply by pluiedev** [reply]
VSCode appears to just embed xterm.js as their integrated terminal, which is licensed under MIT so we can look at how they've implemented it. One difference is that their implementation is entirely in JavaScript with CPU code, while we have to translate that into GPU code (although I doubt it'd be too difficult.)

Here's their implementation, which is quite complicated but I think it would translate well to GLSL:
https://github.com/xtermjs/xterm.js/blob/2042bb85023714e55c0c2e986b5000e33b17c414/src/common/Color.ts#L288-L313

**Reply by pluiedev** [reply]
As for APCA, we should hold off its adoption for multiple reasons. It has very stringent licensing and compliance requirements (see https://github.com/Myndex/SAPC-APCA/blob/master/LICENSE.md) and it specifically worries me in that its license does not allow for the following use cases (emphasis mine):

> Said license excludes other use cases **not related to web content**. Prohibited uses include medical, clinical evaluation, human safety related, aerospace, transportation, military applications, and uses **which are not specific to web-based content** presented on self-illuminated displays or devices.

Ghostty is not a web application and would technically fall into this category. Now, I'm not a lawyer and I'm sure if we ask nicely we'd be explicitly granted permission to use APCA, but this is a legal gray zone that I'm not interested in exploring. *If* APCA (or one of its variations) ever becomes a web open standard, then I'd imagine the licensing requirements would be greatly relaxed.

It should be noted that is also only a potential ***candidate*** for WCAG 3 (which does not yet exist) and it is possible that WCAG 3 would use a modified version of APCA, or a different algorithm entirely. (In fact, the definition for "contrast ratio test" for WCAG 3's working draft as of time of writing is [left empty](https://www.w3.org/TR/wcag-3.0/#dfn-contrast-ratio-test), and nowhere is APCA mentioned in the rest of the document.) Until that open standard has been published, finalized and recommended by W3, we should continue to use WCAG 2 as it is still the current recommendation.

From a technical perspective, I'm not an expert at color science but the formulae and constant definitions appear to be significantly more complex than WCAG 2, and the fact that the sole reference implementation is in (IMO) *very* hard-to-read JavaScript doesn't help either. It also only has predefined values for sRGB, which might make the macOS side of the story more difficult as the native colorspace there is Display P3 (though, I'm not sure how big of an impact that would have — cc @qwerasd205), and I don't exactly know how to get the equivalent constants for other color spaces. These all obstruct its adoption significantly for us as maintainers.

**Comment by Myndex** [comment]
You are expressly granted permission. APCA is intended as a free to use perceptual model to improve accessibility. The license is being worked on to make it easier to adopt as well.


Andrew Somers
_Director of Research_       
_Inclusive Vision Systems Inc._        
https://readtech.org/ARC/
